devito/ir/equations/algorithms.py

```diff
@@ -147,3 +150,51 @@ def lower_exprs(expressions, **kwargs):
     else:
         assert len(processed) == 1
         return processed.pop()
+
+
+def separate_dimensions(expressions):
+    """
+    Rename Dimensions with clashing names within expressions.
+    """
+    resolutions = {}
+    count = {}  # Keep track of increments on dim names
+    processed = []
+    for e in expressions:
+        # Just want dimensions which appear in the expression
+        # Dimensions in indices
+        dims = set().union(*[i.function.dimensions
+                             for i in retrieve_indexed(e)])
+        # Dimensions in conditions and ConditionalDimension parents
+        dims = dims.union(*[pull_dims(d.condition, flag=False)
+                            for d in dims if d.is_Conditional
+                            and d.condition is not None],
+                          set(d.parent for d in dims if d.is_Conditional))
+        # Sort for groupby
+        dims = sorted(dims, key=lambda x: x.name)
+
+        # TODO: Needs to check for dimensions in factors too
+
+        # Group dimensions by matching names
+        groups = tuple(tuple(g) for n, g in groupby(dims, key=lambda x: x.name))
+        clashes = tuple(g for g in groups if len(g) > 1)
+
+        subs = {}
+        for c in clashes:
+            # Preferentially rename dims that aren't root dims
+            ddims = tuple(d for d in c if not d.is_Root)
+
+            for d in ddims:
+                if d in resolutions.keys():
+                    subs[d] = resolutions[d]
+                else:
+                    try:
+                        subs[d] = d._rebuild('%s%s' % (d.name, count[d.name]))
```

Comment:
use the `sregistry` instead of `count`

You can plumb it down the compiler pass from the caller site -- it's one of the kwargs